deepmd/train/trainer_mt.py

```diff
@@ -0,0 +1,713 @@
+                prec = GLOBAL_ENER_FLOAT_PRECISION
+            self.place_holders[kk] = tf.placeholder(prec, [None], name = 't_' + kk)
+            self.place_holders['find_'+kk] = tf.placeholder(tf.float32, name = 't_find_' + kk)
+
+        self.place_holders['type']      = tf.placeholder(tf.int32,   [None], name='t_type')
+        self.place_holders['natoms_vec']        = tf.placeholder(tf.int32,   [data.get_ntypes()+2], name='t_natoms')
+        self.place_holders['default_mesh']      = tf.placeholder(tf.int32,   [None], name='t_mesh')
+        self.place_holders['is_training']       = tf.placeholder(tf.bool)
+        
+        for model_name in self.model_dict.keys():
+            sub_model = self.model_dict[model_name]
+            suffix_dict = self.model_component[model_name]
+            suffix_dict['type_embed'] = 'type_embedding'
+            tmp_model_pred\
+                = sub_model.build (self.place_holders['coord'], 
+                                self.place_holders['type'], 
+                                self.place_holders['natoms_vec'], 
+                                self.place_holders['box'], 
+                                self.place_holders['default_mesh'],
+                                self.place_holders,
+                                suffix = suffix_dict, 
+                                reuse = tf.AUTO_REUSE)
+
+        
+            sub_loss = self.loss_dict[model_name] # model name should be the same as fitting and loss 
+            #if sub_loss.get_name() == data_name:
+            tmp_l2_l, tmp_l2_more\
+                    = sub_loss.build (self.learning_rate_dict[model_name],
+                               self.place_holders['natoms_vec'], 
+                               tmp_model_pred,
+                               self.place_holders,
+                               suffix = model_name)
+            self.l2_l_dict[model_name] = tmp_l2_l
+            self.l2_more_dict[model_name] = tmp_l2_more
+
+        log.info("built network")
+
+    def _build_training(self,data):
+        data_dict,data_name = data.get_data_dict() 
+        trainable_variables = tf.trainable_variables()
+        self.optimizer_dict = {}
+        for loss_name in self.loss_dict.keys():
+            sub_loss = self.loss_dict[loss_name]
+            optimizer = tf.train.AdamOptimizer(learning_rate = self.learning_rate_dict[loss_name])
+
+            if self.run_opt.is_distrib :
+                optimizer = tf.train.SyncReplicasOptimizer(
+                    optimizer,
+                    replicas_to_aggregate = self.run_opt.cluster_spec.num_tasks("worker"),
+                    total_num_replicas = self.run_opt.cluster_spec.num_tasks("worker"),
+                    name = "sync_replicas")
+                self.sync_replicas_hook = optimizer.make_session_run_hook(self.run_opt.is_chief)   
+            self.optimizer_dict[loss_name] = optimizer  
+        grads = tf.gradients(self.l2_l_dict[data_name], trainable_variables)
```

Comment:
The `data_name` comes from the first sampled batch and will be frozen into the TensorFlow graph.
Thus, each step it will only execute the sub graph of the loss minimizing defined on `data_name`.
In other words, either failures of session runs occur or only 1 target is trained on unexpected system frames.